README.md
# Script to extract sentences from Wikipedia articles which contain a link

##### Input: Wikipedia dump 
##### Output: File with triples as following: 
\<http://dbpedia.org/resource/Cyrano_de_Bergerac> \<http://dbpedia.org/resource/Sannois> "He died over a year later on July 28, 1655, aged 36, at the house of his cousin, Pierre De Cyrano, in [[Sannois]]." \
*Subject:* Link to the current article in dbpedia \
*Predicate:* Page Link \
*Object:* Sentence with link in brackets 

## Usage:
**CHANGE TO FOLDER \wikiextractor**
    
    usage: WikiExtractor.py [-h] [-o OUTPUT] -l input

    Process wikidumps to extract all links from the articles and the according sentences
    (There is much more possible with the WikiExtractor, but this is enough for this use)

    positional arguments:
      input                 XML wiki dump file

    optional arguments:
      -h, --help            show this help message and exit
      -o OUTPUT, --output OUTPUT
                            directory for RESULTFILE
      
## Files

wikiextractor/WikiExtractor.py : cleans the subfiles for further processing 

Extract_Sentences.py : Extracts the sentences and its links from the articles and creates result sub files for each directory created by the WikiExtractor (wiki_sum.txt)

Post_WikiExtractor.py : Sums all files up to one RESULT FILE (wiki_triples.txt)

### Running on Windows
To run the script on Windows the encoding needs to be changed at following locations: \
*WikiExtractor.py:* Line 2843 change encoding to:input = fileinput.FileInput(input_file, openhook=fileinput.hook_encoded('cp65001')) \
*Extract_Sentence.py:* Line 19 add following encoding: with open(INPUT_PATH, encoding='cp65001') as wiki_f: \
*Extract_Sentence.py:* Line 108 add following encoding: resulting_file= open(resulting_path+ '/res' + str(now.month) + str(now.day) + '.txt', "w+", encoding='cp65001') \

 





wikiextractor/ChangeLog
2017-01-15  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (process_dump): use text_type to decide whether
	to use decode('utf-8').)

2016-10-29  Giuseppe Attardi  <attardi@di.unipi.it>

	* setup.py: use scripts instead of console_scripts.

	* WikiExtractor.py (pages_from): handle self closing <text />.

2016-08-31  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (findMatchingBraces): fixed ambiguous cases
	like {{{!}} {{!}}}

2016-08-30  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor.wiki2text): drop templates before
	dropping tables.
	(Extractor.templateParams): fixed RE for separating var from value.

	(main): removed option xml_namespaces, since we extract only
	articles, i.e. those with namespace 0.

	(Extractor.clean): removed option --escapedoc, use --html instead.

	* WikiExtractor.py (Extractor.transform): Keep <nowiki> or else
	further expansion will be triggered.

2016-08-29  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (callParserFunction): reworked order of
	evaluation of template parameters, according to
	   https://www.mediawiki.org/wiki/Help:Templates#Order_of_evaluation
        Template parameters are fully evaluated before they are passed to the template.
	The branching parser function are an exception, since evaluation
	depends on values tested.

	* WikiExtractor.py (callParserFunction): reworked handling of
	parser function.
	Handle expansion of:
	   {{numero romano|1999}}
	where Template:Numero romano = {{{{{|safesubst:}}}#invoke:Numero romano|main}}
	and parameter body has no arguments, by getting them from frame.
	Instead expansion of:
	   {{Str endswith|1999|a.C.}}
	where Template:Str endswith = {{#ifeq:{{#invoke:String|sub|s={{{1|}}}| -{{#invoke:String|len|s={{{2}}}}} |ignore_errors=true}}|{{{2|}}}|sì}}
	uses parameters in template.

	* WikiExtractor.py (Extractor.extract): Adds an h1 tag for title if --html is
	passed. Patch contributed by okb1100 on GitHub.

	* WikiExtractor.py (modules): added partial emulation of modules String and Roman.
	)
2016-08-26  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (nowiki): handle <nowiki>...</nowiki>.

2016-08-22  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor.extract): fixed UTF-8 encoding with
	option -a.

2016-08-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor): fixed misspelled keepLists.

2016-08-11  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extract): made escape_doc a class variable.

2016-06-20  Seth Cleveland <scleveland@turnitin.com>

	* WikiExtractor.py: Fix string encoding error for python 2 and log
	extract exceptions.

2016-06-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py: support for Python 3 by orangain@gmail.com

2016-03-23  Seth Cleveland <scleveland@turnitin.com>

	* WikiExtractor.py: add filtering options -- min text length, xml
	namespace, and disambig pages. Added option to print the revision
	id with each document.

2016-03-23  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (compact): properly emit section headers when Extractor.keepSections.

2016-03-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (main): restored option --sections.

2016-03-12  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (findBalanced): fix to bad single tuple
	parameter to findBalanced.

2016-03-06  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (process_dump): handle templates with stdin.

2016-02-20  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (reduce_process): tell mapper to wait when
	spool gets too long.

2016-02-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (extract_process): use out.truncate() instread
	of out.close()

2016-02-15  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor.clean): turned into method.
	(process_dump): control spool length to avoid filing up memory.

2016-02-14  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (load_templates): save also <id> in templates.
	(pages_from): common iterator for extracting pages from dump, used
	both for analyzing pages, templates and single articles.

2016-02-13  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (reduce_process): close file here.

	* extractPage.py (process_data): allow range of ids.

2016-02-12  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (reduce_process): moved here creation of OutputSplitter.
	(compact): Extractor.keepLists allows preserving lists in output.
	(main): added new option --lists for preserving lists in output.
	(compact): rest lislLevel entering new section.
	(ignoredTags): removed 'div' from here, since it is in discardedTags.
	(ignoredTags): moved 'sub' and 'sup' to discardedTags.

2016-02-11  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (if_empty): added implementation of Lua module
	for "Template:If empty".
	(compact): preserve lists also in text mode.
	(process_dump): log ID and title of pages.

2016-02-10  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (process_dump): detect templates by element
	<ns>10</ns> rather than by colon in title.
	(Extractor.templateParams): restore space in RE [^= ]:
	m = re.match(' *([^= ]*?) *=(.*)', param, re.DOTALL)

2016-02-04  Giuseppe Attardi  <attardi@di.unipi.it>

	* cirrus-extract.py: added.
	* README.md: added mention of Cirrus extract.

2015-11-20  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (makeExternalLink): fixed.
	(main): dropped confusing option --section.

2015-10-29  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py: /usr/bin/env

2015-09-29  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py: fixed argparse default.

Semantic_WikiLinks.py
import Extract_Sentences
import Post_WikiExtractor
import os
import time
from multiprocessing import Pool
import shutil

start = time.time()

def main(output_path, nb_processes):
    '''
    The WikiExtractor splits the articles in thousands of files, which will be first used to parallelize
    (Pool multiprocessing) the creation of the triples for each article. In Post_WikiExtractor.sum_up() the result file
    will be summed up to one result file /wiki_triples.txt
    :param output_path: output path, which is given via command input
    :param nb_processes: number of processes to parallelize
    '''
    dir_data = []
    seen = set(dir_data)
    for root, dirs, files in os.walk(output_path):
        if files:
            complete_sub_wiki = open(root + '/wiki_sum.txt', 'wb')
            for subfile in files:
                subfile_dir = os.path.join(root, subfile)
                open_subfile = open(subfile_dir, 'rb')
                shutil.copyfileobj(open_subfile, complete_sub_wiki)
                if root not in seen:
                    seen.add(root)
                    dir_data.append([root + '/wiki_sum.txt', root])
    # use Pool for multiprocessing
    p = Pool(processes=nb_processes)
    p.map(Extract_Sentences.result_file, dir_data)
    end_multi = time.time()

    # sum files up to result file
    ROOTDIR = output_path
    Post_WikiExtractor.sum_up(ROOTDIR, output_path)

    # print('3 ALL FILES SUMMED UP in {}'.format(time.time() - start))
    end = time.time()

    # time tracking
    print('Time tracking: complete process {}'.format(end - start))
    print('Time tracking: Summing up {}'.format(end - end_multi))

if __name__ == '__main__':
    main()

Extract_Sentences.py
import re
import datetime
import time
from nltk.tokenize import sent_tokenize
import nltk
import urllib

# this script extracts all links and the according sentences from the wikidump files (after parsing with WikiExtractor.py
# input: parsed wiki file
# output: result file with triples like: subject as article name, predicate as entity, object as the according sentence

amount_articles = 0

# input --> result_file from wikiExtractor
# wiki articles, splitted by <doc id='' url='' title='' ></doc>
def open_wiki_files(INPUT_PATH):
    article_list = []
    article = ""
    with open(INPUT_PATH) as wiki_f: #, encoding='cp65001'
        wiki_file_line = wiki_f.readline()
        while (wiki_file_line):
            article += wiki_file_line
            if '</doc>' in wiki_file_line:
                article_list.append(article)
                article = ""
            wiki_file_line = wiki_f.readline()
    return article_list, len(article_list)

def write_file(title, entity, sentence, file):
    title = title.replace(' ', '_')
    entity = entity.replace(' ', '_')
    file.write(
            '<http://dbpedia.org/resource/' + title + '> ' + '<http://dbpedia.org/resource/' + entity + '> ' + '\"' + sentence + '\" \n')


def extract_header(article):
    re_article_header = re.search(r'<doc .*?>', article)
    article_header = re_article_header.group()

    re_article_url = re.search(r'url="(.*?)"', article)
    article_url = re_article_url.group()
    url = article_url.split('"')
    return article_header


def extract_title(article):
    re_article_title = re.search(r'title=".*?"', article)
    article_title = re_article_title.group()
    title = article_title.split('"')
    return title[1]


def parse_entity(entity):
    entity = urllib.parse.unquote(entity)
    entity = entity.replace('&amp;', '&')
    return entity


def extract_sentence(sentence):
    '''
    This function replaces the links with its entity and gives back two lists: one containing the entities for the
    resulting triple and the other list containing the matching sentence.
    :param sentence: the raw sentence containing the raw links indicated by <a href= >
    :return: one list containing the entities of the sentence, the other list containing the sentence with the according
    entity in brackets
    '''
    sentence = sentence.replace('\n', '')
    sentence_list = []
    entity_list = []
    sentence_link = re.findall(r'<a href=.*?</a>', sentence)
    for link_element in sentence_link:
        # extract entity from links and replace every entity in the sentence
        link_entity = link_element.replace('>', '<').split('<')[2]
        sentence = sentence.replace(link_element, link_entity)
    for link_element in sentence_link:
        # mark iteratively each entity in the sentence as entity and save sentence + entity
        link_entity_obj = link_element.replace('>', '<').split('<')[2]
        sentence_entity = sentence.replace(link_entity_obj, '[[' + link_entity_obj + ']]')
        link_entity_pred = link_element.split('"')[1]  # <--------------
        link_entity_pred = parse_entity(link_entity_pred)  # <--------------
        sentence_list.append(sentence_entity)
        entity_list.append(link_entity_pred)
    return entity_list, sentence_list


def result_file(path):
    '''
    splits file into its articles and safes into a list (def open_wiki_files()). By iterating through the articles
    first all links are extracted via regex and replaced (since often periods are in the links, it was a source of
    mistakes for splitting it into sentences). After splitting the article into its sentences via
    nltk library (https://www.nltk.org/), the wildcards for the links are replaced by the links again. Finally the
    according sentences and the including entities are extracted and saved into the result_file.
    :param path: array of input and output directory
    :return: a file with all sentences which contain a link. The sentences are saved in triples:
    the article link as subject, the link (entity) as predicate, and the sentence which contained the entity as object
    e.g. <http://dbpedia.org/resource/Cyrano_de_Bergerac> <http://dbpedia.org/resource/Sannois> "He died over a year
    later on July 28, 1655, aged 36, at the house of his cousin, Pierre De Cyrano, in [[Sannois]]."
    '''
    INPUT_PATH = path[0]
    resulting_path = path[1]
    start = time.time()
    # add e.g. as abbreviation to set
    extra_abbreviation = ['e.g', 'co', 'st', 'mr']
    sentence_tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')
    sentence_tokenizer._params.abbrev_types.update(extra_abbreviation)

    now = datetime.datetime.now()
    resulting_file= open(resulting_path+ '/res' + str(now.month) + str(now.day) + '.txt', "w+") #, encoding='cp65001'

    articles = open_wiki_files(INPUT_PATH)
    amount_articles =articles[1]
    for article in articles[0]:
        i = 0
        nb_entities = 0
        header = extract_header(article)
        title = extract_title(article)
        article = article.replace(header, '').replace(title, '', 1)
        re_links = re.findall(r'<a href=.*?</a>', article)
        for link in re_links:
            article = article.replace(link, '[[entity]]')
        article_in_sentences = sent_tokenize(article) # split into sentence with nltk library
        for raw_sentence in article_in_sentences:
            while '[[entity]]' in raw_sentence:
                raw_sentence = raw_sentence.replace('[[entity]]', re_links[i], 1)
                i+=1
            elements = extract_sentence(raw_sentence)
            nb_entities = len(elements[0])
            for e in range(nb_entities):
                write_file(title, elements[0][e], elements[1][e], resulting_file)

    end = time.time()

    print('AMOUNT OF ARTICLES {}'.format(amount_articles))
    print('--- TIME {}'.format(end - start))
